pkg/ee/validation/machine/resourcequota.go

```diff
@@ -0,0 +1,155 @@
+                         Version 1.0 ("KERO-1.0”)
+                     Copyright © 2022 Kubermatic GmbH
+
+   1.	You may only view, read and display for studying purposes the source
+      code of the software licensed under this license, and, to the extent
+      explicitly provided under this license, the binary code.
+   2.	Any use of the software which exceeds the foregoing right, including,
+      without limitation, its execution, compilation, copying, modification
+      and distribution, is expressly prohibited.
+   3.	THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND,
+      EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+      MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+      IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+      CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+      TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+      SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+   END OF TERMS AND CONDITIONS
+*/
+
+package machine
+
+import (
+	"context"
+	"fmt"
+
+	"go.uber.org/zap"
+
+	"github.com/kubermatic/machine-controller/pkg/apis/cluster/v1alpha1"
+	"github.com/kubermatic/machine-controller/pkg/providerconfig/types"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ValidateQuota validates if the requested Machine resource consumption fits in the quota of the clusters project.
+func ValidateQuota(ctx context.Context, log *zap.SugaredLogger, seedClient ctrlruntimeclient.Client, machine *v1alpha1.Machine) error {
+	config, err := types.GetConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return fmt.Errorf("failed to read machine.spec.providerSpec: %w", err)
+	}
+
+	// TODO add all providers
+	var quotaReq *ResourceQuota
+	switch config.CloudProvider {
+	// add this fake for test and so further code is reachable until more providers are implemented
+	case types.CloudProviderFake:
+		quotaReq, err = getFakeQuotaRequest(config)
+		if err != nil {
+			return fmt.Errorf("error getting fake quota req: %w", err)
+		}
+	default:
+		// TODO skip for now, when all providers are added, throw error
+		log.Debugf("provider %q not supported", config.CloudProvider)
+		return nil
+	}
+
+	// TODO Get quota and usage from ResourceQuota when its implemented
+	quota, currentUsage, err := getResourceQuota()
+	if err != nil {
+		return fmt.Errorf("failed to get resource quota: %w", err)
+	}
+
+	// add requested resources to current usage and compare
+	combinedUsage := NewResourceQuota(currentUsage.cpu, currentUsage.mem, currentUsage.storage)
+	combinedUsage.Cpu().Add(*quotaReq.Cpu())
+	combinedUsage.Memory().Add(*quotaReq.Memory())
+	combinedUsage.Storage().Add(*quotaReq.Storage())
+
+	if quota.Cpu().Cmp(*combinedUsage.Cpu()) < 0 {
+		log.Debugf("requested CPU %q would exceed current quota (quota/used %q/%q)",
```

Comment:
I love readable log messages, but should we use structured logging here? Honest question. "No." is a valid answer.

Reply:
Changed to structured, looks better